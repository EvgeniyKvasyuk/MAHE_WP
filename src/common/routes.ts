export const ROUTES = {
  MAIN_PAGE: '/',
  APPOINTMENTS: {
    ROOT: '/appointments',
    SCHEDULE_APPOINTMENT: '/schedule',
    VIEW_APPOINTMENTS: '/view',
  },
  MY_HEALTH: {
    ROOT: '/my_health',
    MY_HEALTH_DASHBOARD: '/My Health Dashboard',
  },
  MESSAGES: '/messages',
  PROFILE: '/profile',
  PRESCRIPTIONS: '/prescriptions',
  ARTICLE: '/article',
};
