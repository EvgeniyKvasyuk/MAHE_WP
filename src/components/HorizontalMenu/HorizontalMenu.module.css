@import '_z-index.css';
@import '_fonts.css';
@import '_sizes.css';

.horizontal-menu {
  z-index: $z-index-main-menu;
  display: flex;
  flex-direction: row;
  padding-top: 0;
  padding-bottom: 0;
  font-size: $medium-font-size;

  &__tab {
    margin-top: 1rem;
    padding-top: 0.5rem;
  }
}
