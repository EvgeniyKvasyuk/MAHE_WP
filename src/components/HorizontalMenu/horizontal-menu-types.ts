export type SubMenuItemType = {
  to: string;
  label: string;
};

export type TabProps = {
  id: number;
  to: string;
  label: string;
  subMenuItems?: Array<SubMenuItemType>;
};
