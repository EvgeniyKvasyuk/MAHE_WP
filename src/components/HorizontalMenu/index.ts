export { HorizontalMenu } from './HorizontalMenu';
export type { TabProps } from './horizontal-menu-types';
