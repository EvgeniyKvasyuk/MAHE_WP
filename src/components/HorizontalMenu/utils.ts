// TODO-Vladislav write tests on utils
import { TabProps } from './horizontal-menu-types';

export const getMenuId = (id: number) => `sub-menu-${id}`;

export const getTabById = (tabId: number, tabList: Array<TabProps>) => tabList.findIndex(({ id }) => id === tabId);

export const isElementInPopup = (element: HTMLElement, anchorEl: HTMLElement, id: number | null): boolean => {
  if (!id || !element) {
    return false;
  }
  const menuElement = document.getElementById(getMenuId(id));

  return !!menuElement?.contains?.(element);
};

const matchPath = (pathName: string, to: string): boolean => {
  const splittedPathName = pathName.split('/').filter(Boolean);

  return splittedPathName.includes(to.replace('/', ''));
};

export const getInitialTabId = (pathname: string, tabList: Array<TabProps>) => {
  const tab = tabList.find(({ to }) => matchPath(pathname, to));

  return tab?.id;
};
