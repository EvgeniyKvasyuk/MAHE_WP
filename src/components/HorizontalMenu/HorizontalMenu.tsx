import React, { useCallback, useMemo, useState } from 'react';

import classNames from 'classnames/bind';
import { useLocation, useNavigate } from 'react-router-dom';

import { isTouchDevice } from '@common/utils';
import { AppBar } from '@components/AppBar';
import { Tab, Tabs } from '@components/Tabs';
import { useFlippers } from '@modules/flippers';

import styles from './HorizontalMenu.module.css';
import { SubMenu } from './components';
import { useOpenOnHover } from './hooks';
import { TabProps } from './horizontal-menu-types';
import { getInitialTabId, getTabById } from './utils';

const cn = classNames.bind(styles);

export interface HorizontalMenuProps {
  menuItemList: Array<TabProps>;
  className?: string;
}

export function HorizontalMenu({ menuItemList, className }: HorizontalMenuProps) {
  const location = useLocation();
  const navigate = useNavigate();
  const getFlipperValue = useFlippers();

  const visibleMenuItems = useMemo(() => {
    return menuItemList.filter((menuItem) => (menuItem.flipper ? !getFlipperValue(menuItem.flipper) : true));
  }, [menuItemList, getFlipperValue]);

  const [selectedTab, setSelectedTab] = useState(() => getInitialTabId(location.pathname, visibleMenuItems));

  const { closeMenu, handleTabMouseLeave, handleTouchStart, handleTabMouseOver, hoveredTabId, menuAnchorEl } =
    useOpenOnHover();

  const selectTabAndNavigate = useCallback(
    (tabId: number, path: string) => {
      if (path) {
        setSelectedTab(tabId);
        navigate(path);
      }
    },
    [navigate],
  );

  const handleTabChange = useCallback(
    (event, newId) => {
      const tabIndex = getTabById(newId, visibleMenuItems);
      const tab = visibleMenuItems[tabIndex];

      if (tab.to) {
        selectTabAndNavigate(newId, tab.to);
        if (!isTouchDevice()) {
          closeMenu();
        }
      }
    },
    [visibleMenuItems, selectTabAndNavigate, closeMenu],
  );

  const menuItemsWithSubMenu = useMemo(
    () => visibleMenuItems.filter((menuItem) => menuItem.subMenuItems?.length),
    [visibleMenuItems],
  );

  return (
    <AppBar className={cn('horizontal-menu', className)}>
      <Tabs onChange={handleTabChange} value={selectedTab}>
        {visibleMenuItems.map((tab) => (
          <Tab
            key={tab.id}
            value={tab.id}
            label={tab.label}
            onMouseOver={tab.subMenuItems ? handleTabMouseOver(tab.id) : undefined}
            onMouseLeave={tab.subMenuItems ? handleTabMouseLeave : undefined}
            onTouchStart={tab.subMenuItems ? handleTouchStart(tab.id) : undefined}
            selected={tab.id === selectedTab}
            hovered={tab.id === hoveredTabId}
            className={cn('horizontal-menu__tab')}
          />
        ))}
      </Tabs>

      <SubMenu
        closeMenu={closeMenu}
        hoveredTabId={hoveredTabId}
        menuAnchorEl={menuAnchorEl}
        menuItemsWithSubMenu={menuItemsWithSubMenu}
        selectTabAndNavigate={selectTabAndNavigate}
      />
    </AppBar>
  );
}
