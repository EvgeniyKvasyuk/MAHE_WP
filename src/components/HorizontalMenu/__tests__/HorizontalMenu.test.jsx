import React from 'react';

import { ROUTES } from '@common/routes';
import { render } from '@tests/test-utils';

import { HorizontalMenu } from '../HorizontalMenu';

const APPOINTMENTS_SUB_MENU_ITEMS = [
  {
    label: 'Schedule Appointment',
    to: ROUTES.APPOINTMENTS.SCHEDULE_APPOINTMENT,
  },
  {
    label: 'View Appointments',
    to: ROUTES.APPOINTMENTS.VIEW_APPOINTMENTS,
  },
];

export const TAB_LIST = [
  {
    id: 0,
    label: 'Home',
    to: ROUTES.MAIN_PAGE,
  },
  {
    id: 1,
    label: 'Appointments',
    to: ROUTES.APPOINTMENTS.ROOT,
    subMenuItems: APPOINTMENTS_SUB_MENU_ITEMS,
  },
  {
    id: 2,
    label: 'My Health',
    to: ROUTES.MY_HEALTH.ROOT,
  },
];

test('<HorizontalMenu> snapshot', () => {
  const component = render(<HorizontalMenu menuItemList={TAB_LIST} />);

  expect(component).toMatchSnapshot();
});
