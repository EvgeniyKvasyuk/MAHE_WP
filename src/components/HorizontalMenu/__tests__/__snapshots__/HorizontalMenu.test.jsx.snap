// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HorizontalMenu> snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header
        class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic app-bar horizontal-menu css-hip9hq-MuiPaper-root-MuiAppBar-root"
      >
        <div
          class="app-bar__content"
        >
          <div
            class="MuiTabs-root tabs css-1ujnqem-MuiTabs-root"
          >
            <div
              class="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
              style="overflow: hidden; margin-bottom: 0px;"
            >
              <div
                class="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
                role="tablist"
              >
                <button
                  class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit tab horizontal-menu__tab css-cjctlb-MuiButtonBase-root-MuiTab-root"
                  role="tab"
                  tabindex="-1"
                  type="button"
                >
                  Home
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
                <button
                  class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit tab horizontal-menu__tab css-cjctlb-MuiButtonBase-root-MuiTab-root"
                  role="tab"
                  tabindex="-1"
                  type="button"
                >
                  Appointments
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
                <button
                  class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit tab horizontal-menu__tab css-cjctlb-MuiButtonBase-root-MuiTab-root"
                  role="tab"
                  tabindex="-1"
                  type="button"
                >
                  My Health
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
              <span
                class="MuiTabs-indicator tabs__indicator css-1aquho2-MuiTabs-indicator"
                style="left: 0px; width: 0px;"
              />
            </div>
          </div>
        </div>
      </header>
    </div>
  </body>,
  "container": <div>
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic app-bar horizontal-menu css-hip9hq-MuiPaper-root-MuiAppBar-root"
    >
      <div
        class="app-bar__content"
      >
        <div
          class="MuiTabs-root tabs css-1ujnqem-MuiTabs-root"
        >
          <div
            class="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
            style="overflow: hidden; margin-bottom: 0px;"
          >
            <div
              class="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
              role="tablist"
            >
              <button
                class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit tab horizontal-menu__tab css-cjctlb-MuiButtonBase-root-MuiTab-root"
                role="tab"
                tabindex="-1"
                type="button"
              >
                Home
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
              <button
                class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit tab horizontal-menu__tab css-cjctlb-MuiButtonBase-root-MuiTab-root"
                role="tab"
                tabindex="-1"
                type="button"
              >
                Appointments
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
              <button
                class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit tab horizontal-menu__tab css-cjctlb-MuiButtonBase-root-MuiTab-root"
                role="tab"
                tabindex="-1"
                type="button"
              >
                My Health
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
            <span
              class="MuiTabs-indicator tabs__indicator css-1aquho2-MuiTabs-indicator"
              style="left: 0px; width: 0px;"
            />
          </div>
        </div>
      </div>
    </header>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
